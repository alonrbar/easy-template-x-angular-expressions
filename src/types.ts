
export interface IMap<T> {
    [key: string]: T;
}
