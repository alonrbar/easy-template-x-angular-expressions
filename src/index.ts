export * from './createResolver';
export * from './pathFilter';
export * from './resolveError';
export * from './resolver';
export * from './resolverOptions';
export * from './types';
export * from './utils';
