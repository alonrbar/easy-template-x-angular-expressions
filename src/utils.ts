export function isNumber(value: unknown): value is number {
    return Number.isFinite(value);
}

// eslint-disable-next-line @typescript-eslint/ban-types
export function isObject(value: unknown): value is object {
    return value !== null && value !== undefined && typeof value === 'object';
}

/**
 * Naive string escaping.
 */
export function strEscape(str: string): string {
    return str.replace(/\"/g, '\\"');
}
