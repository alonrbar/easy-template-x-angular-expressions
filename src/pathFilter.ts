import { isNumber } from './utils';

export type PathFilter = (part: string | number) => boolean;

const simpleToken = "[a-zA-z_]\\w*";

const bracketToken = `\\[("[^"]+"|'[^']+'|\\d+)\\]`;

const token = `((${simpleToken})|(${bracketToken}))`;

const defaultFilterRegex = new RegExp(`^\\s*${token}(\\.${simpleToken}|${bracketToken})*\\s*$`);

/**
 * The default path filters returns true for js object paths and false for other
 * js expressions.
 */
export const defaultPathFilter : PathFilter = (part: string | number) => {
    if (isNumber(part)) {
        return true;
    }
    return defaultFilterRegex.test(part);
};
