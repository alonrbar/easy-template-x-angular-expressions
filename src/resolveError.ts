import { strEscape } from './utils';

export class ResolveError extends Error {

    constructor(
        public readonly expression: string,
        public readonly path: string[],
        public readonly innerError: Error
    ) {
        super(`Failed to resolve expression "${strEscape(expression)}" (path: ${printPath(path)}). Inner error: ${innerError?.message}.`);

        // typescript hack: https://github.com/Microsoft/TypeScript-wiki/blob/master/Breaking-Changes.md#extending-built-ins-like-error-array-and-map-may-no-longer-work
        Object.setPrototypeOf(this, ResolveError.prototype);
    }
}

function printPath(path: string[]): string {
    const strPath = (path || [])
        .map(p => `"${strEscape(p ?? '')}"`)
        .join(", ");
    return `[ ${strPath} ]`;
}
