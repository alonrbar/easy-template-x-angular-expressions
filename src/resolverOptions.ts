import { defaultPathFilter, PathFilter } from './pathFilter';
import { IMap } from './types';

export type AngularFilter = (value: any) => any;

export class ResolverOptions {
    /**
     * If set to a non-empty string the resolver will require the tag name to
     * have this prefix (ignoring leading whitespace).
     *
     * Default: false
     */
    public requiredPrefix?: string | false = false;
    /**
     * The path filter is used to determine which path parts to consider when
     * traversing the data object to construct the expression scope. Path parts
     * that returns `false` are skipped and the traversal moves to the next
     * part.
     */
    public pathFilter?: PathFilter = defaultPathFilter;
    /**
     * Setting this option to `true` instructs the resolver to fallback to the default `easy-template-x` data resolver.
     * Setting it to `false` will return `undefined` instead.
     *
     * The fallback is used in the following cases:
     *
     *   1. Empty path.
     *   2. Numeric path parts (loop iteration index).
     *   3. When the `requiredPrefix` is specified and the tag name does not meet the requirement.
     *
     * Default: true
     */
    public defaultFallback?= true;
    /**
     * Map of Angular filters.
     *
     * **Notice**: The filters are global. That is, filters of one Resolver may
     * override filters of another Resolver.
     */
    public angularFilters?: IMap<AngularFilter> = {};

    constructor(initial?: Partial<ResolverOptions>) {
        Object.assign(this, initial);
        if (this.requiredPrefix && this.requiredPrefix.trim() != this.requiredPrefix) {
            throw new Error("requirePrefix cannot contain leading or trailing whitespace");
        }
    }
}
