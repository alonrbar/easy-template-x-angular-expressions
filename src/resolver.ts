import { ScopeData } from "easy-template-x";
import { ResolveError } from "./resolveError";
import { ResolverOptions } from "./resolverOptions";
import { isNumber, isObject } from "./utils";
import type { ScopeDataArgs, ScopeDataResolver, TemplateContent, TemplateData } from 'easy-template-x';
const expressions = require('angular-expressions');
const getProp = require("lodash.get");

const undefinedResolver: ScopeDataResolver = () => undefined;

export class AngularResolver {

    private readonly options: ResolverOptions;
    private readonly fallback: ScopeDataResolver;

    constructor(options?: ResolverOptions) {
        this.options = new ResolverOptions(options);
        this.fallback = this.options.defaultFallback ? ScopeData.defaultResolver : undefinedResolver;

        // Configure 'angular-expressions' filters.
        for (const key of Object.keys(this.options.angularFilters || {})) {
            expressions.filters[key] = this.options.angularFilters[key];
        }
    }

    public resolve(args: ScopeDataArgs): TemplateContent | TemplateData[] {

        // Fallback on empty paths.
        if (!args.path.length) {
            return this.fallback(args);
        }

        // Fallback on number paths (generated by the loop plugin).
        const lastPart = args.path[args.path.length - 1];
        if (isNumber(lastPart)) {
            return this.fallback(args);
        }

        // Check required prefix.
        let exp = (lastPart?.name || "").trim();
        if (this.options.requiredPrefix && !exp.startsWith(this.options.requiredPrefix)) {
            return this.fallback(args);
        }
        if (this.options.requiredPrefix) {
            exp = exp.substr(this.options.requiredPrefix.length);
        }

        // Flatten the scope.
        const finalScope = Object.assign({}, args.data);
        let curScop: any = finalScope;
        for (const part of args.path) {
            const partIndexer = isNumber(part) ? part : part.name;

            // Apply the path filter.
            if (this.options.pathFilter && !this.options.pathFilter(partIndexer)) {
                continue;
            }
            // If it's not an object don't go deeper.
            curScop = getProp(curScop, partIndexer);
            if (!isObject(curScop)) {
                break;
            }
            Object.assign(finalScope, curScop);
        }

        // Resolve the expression.
        exp = exp.replace(/(’|‘)/g, "'").replace(/(“|”)/g, '"');
        try {
            return expressions.compile(exp)(finalScope);
        } catch (e) {
            throw new ResolveError(exp, args.strPath, e);
        }
    }
}
