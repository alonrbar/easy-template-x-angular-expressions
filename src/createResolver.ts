import { ResolverOptions } from './resolverOptions';
import type { ScopeDataResolver } from 'easy-template-x';
import { AngularResolver } from './resolver';

export function createResolver(options?: ResolverOptions): ScopeDataResolver {
    const resolver = new AngularResolver(options);
    return resolver.resolve.bind(resolver);
}
