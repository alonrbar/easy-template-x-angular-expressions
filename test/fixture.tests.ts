import { TemplateHandler } from 'easy-template-x';
import * as fs from 'fs';
import { createResolver } from 'src';
import { removeDuplicateWhitespace } from './testUtils';

describe('fixture tests', () => {

    it('works as expected', async () => {
        const handler = new TemplateHandler({
            scopeDataResolver: createResolver({
                angularFilters: {
                    upper: (input: string) => (input || "").toUpperCase(),
                    lower: (input: string) => (input || "").toLowerCase()
                }
            })
        });
        const template = fs.readFileSync("./test/fixtures/test.docx");
        const templateText = await handler.getText(template);
        expect(removeDuplicateWhitespace(templateText)).toEqual([
            "{# myLoop}",
            "{# casing == “upper”}",
            "{item.name | upper}",
            "{/}",
            "{# casing == “lower”}",
            "{item.name | lower}",
            "{/}",
            "----",
            "{/}"
        ].join(""));

        const data = {
            "myLoop":[
                {
                    casing: "upper",
                    item: {
                        name: "Bla"
                    }
                },
                {
                    casing: "lower",
                    item: {
                        name: "Some"
                    }
                },
            ]
        };

        const doc = await handler.process(template, data);

        const docText = await handler.getText(doc);
        expect(removeDuplicateWhitespace(docText)).toEqual([
            "BLA",
            "----",
            "some",
            "----",
        ].join(""));

        const docXml = await handler.getXml(doc);
        expect(docXml).toMatchSnapshot();

        // fs.writeFileSync('/temp/angular parser - out.docx', doc);
    });
});
