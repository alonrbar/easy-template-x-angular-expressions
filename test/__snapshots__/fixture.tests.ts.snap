// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fixture tests works as expected 1`] = `<w:document xmlns:wpc="http://schemas.microsoft.com/office/word/2010/wordprocessingCanvas" xmlns:cx="http://schemas.microsoft.com/office/drawing/2014/chartex" xmlns:cx1="http://schemas.microsoft.com/office/drawing/2015/9/8/chartex" xmlns:cx2="http://schemas.microsoft.com/office/drawing/2015/10/21/chartex" xmlns:cx3="http://schemas.microsoft.com/office/drawing/2016/5/9/chartex" xmlns:cx4="http://schemas.microsoft.com/office/drawing/2016/5/10/chartex" xmlns:cx5="http://schemas.microsoft.com/office/drawing/2016/5/11/chartex" xmlns:cx6="http://schemas.microsoft.com/office/drawing/2016/5/12/chartex" xmlns:cx7="http://schemas.microsoft.com/office/drawing/2016/5/13/chartex" xmlns:cx8="http://schemas.microsoft.com/office/drawing/2016/5/14/chartex" xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006" xmlns:aink="http://schemas.microsoft.com/office/drawing/2016/ink" xmlns:am3d="http://schemas.microsoft.com/office/drawing/2017/model3d" xmlns:o="urn:schemas-microsoft-com:office:office" xmlns:r="http://schemas.openxmlformats.org/officeDocument/2006/relationships" xmlns:m="http://schemas.openxmlformats.org/officeDocument/2006/math" xmlns:v="urn:schemas-microsoft-com:vml" xmlns:wp14="http://schemas.microsoft.com/office/word/2010/wordprocessingDrawing" xmlns:wp="http://schemas.openxmlformats.org/drawingml/2006/wordprocessingDrawing" xmlns:w10="urn:schemas-microsoft-com:office:word" xmlns:w="http://schemas.openxmlformats.org/wordprocessingml/2006/main" xmlns:w14="http://schemas.microsoft.com/office/word/2010/wordml" xmlns:w15="http://schemas.microsoft.com/office/word/2012/wordml" xmlns:w16cex="http://schemas.microsoft.com/office/word/2018/wordml/cex" xmlns:w16cid="http://schemas.microsoft.com/office/word/2016/wordml/cid" xmlns:w16="http://schemas.microsoft.com/office/word/2018/wordml" xmlns:w16se="http://schemas.microsoft.com/office/word/2015/wordml/symex" xmlns:wpg="http://schemas.microsoft.com/office/word/2010/wordprocessingGroup" xmlns:wpi="http://schemas.microsoft.com/office/word/2010/wordprocessingInk" xmlns:wne="http://schemas.microsoft.com/office/word/2006/wordml" xmlns:wps="http://schemas.microsoft.com/office/word/2010/wordprocessingShape" mc:Ignorable="w14 w15 w16se w16cid w16 w16cex wp14"><w:body><w:p w14:paraId="2C0EFBFA" w14:textId="302A3746" w:rsidR="0005473B" w:rsidRDefault="007824CF"><w:pPr><w:rPr><w:lang w:val="en-US"/></w:rPr></w:pPr><w:r w:rsidR="00476B45"><w:rPr><w:lang w:val="en-US"/></w:rPr></w:r><w:r><w:rPr><w:lang w:val="en-US"/></w:rPr></w:r><w:r w:rsidR="00695252"><w:rPr><w:lang w:val="en-US"/></w:rPr></w:r><w:r><w:rPr><w:lang w:val="en-US"/></w:rPr></w:r></w:p><w:p w14:paraId="765DFE25" w14:textId="7F8B3443" w:rsidR="007824CF" w:rsidRDefault="007824CF"><w:pPr><w:rPr><w:lang w:val="en-US"/></w:rPr></w:pPr><w:r w:rsidR="00695252"><w:rPr><w:lang w:val="en-US"/></w:rPr></w:r><w:r w:rsidR="00935F47"><w:rPr><w:lang w:val="en-US"/></w:rPr></w:r><w:r><w:rPr><w:lang w:val="en-US"/></w:rPr></w:r><w:r w:rsidR="00695252"><w:rPr><w:lang w:val="en-US"/></w:rPr></w:r><w:r><w:rPr><w:lang w:val="en-US"/></w:rPr></w:r></w:p><w:p w14:paraId="5A15DE06" w14:textId="77777777" w:rsidR="00695252" w:rsidRDefault="00695252"><w:pPr><w:rPr><w:lang w:val="en-US"/></w:rPr></w:pPr><w:r w:rsidRPr="00695252"><w:rPr><w:lang w:val="en-US"/></w:rPr><w:t xml:space="preserve">BLA</w:t></w:r></w:p><w:p w14:paraId="27AC477B" w14:textId="2C552915" w:rsidR="007824CF" w:rsidRDefault="007824CF"><w:pPr><w:rPr><w:lang w:val="en-US"/></w:rPr></w:pPr></w:p><w:p w14:paraId="00F1E240" w14:textId="00295242" w:rsidR="007824CF" w:rsidRDefault="007824CF" w:rsidP="007824CF"><w:pPr><w:rPr><w:lang w:val="en-US"/></w:rPr></w:pPr></w:p><w:p w14:paraId="193F9BCE" w14:textId="39E7FA17" w:rsidR="006E722B" w:rsidRDefault="006E722B" w:rsidP="00F13143"><w:pPr><w:rPr><w:lang w:val="en-US"/></w:rPr></w:pPr><w:r><w:rPr><w:lang w:val="en-US"/></w:rPr><w:t>----</w:t></w:r></w:p><w:p w14:paraId="79276D44" w14:textId="4F382710" w:rsidR="007824CF" w:rsidRPr="007824CF" w:rsidRDefault="007824CF"><w:pPr><w:rPr><w:lang w:val="en-US"/></w:rPr></w:pPr><w:r w:rsidR="00476B45"><w:rPr><w:lang w:val="en-US"/></w:rPr></w:r><w:r><w:rPr><w:lang w:val="en-US"/></w:rPr></w:r><w:r w:rsidR="00695252"><w:rPr><w:lang w:val="en-US"/></w:rPr></w:r><w:r><w:rPr><w:lang w:val="en-US"/></w:rPr></w:r></w:p><w:p w14:paraId="765DFE25" w14:textId="7F8B3443" w:rsidR="007824CF" w:rsidRDefault="007824CF"><w:pPr><w:rPr><w:lang w:val="en-US"/></w:rPr></w:pPr></w:p><w:p w14:paraId="00F1E240" w14:textId="00295242" w:rsidR="007824CF" w:rsidRDefault="007824CF" w:rsidP="007824CF"><w:pPr><w:rPr><w:lang w:val="en-US"/></w:rPr></w:pPr><w:r w:rsidR="00695252"><w:rPr><w:lang w:val="en-US"/></w:rPr></w:r><w:r w:rsidR="00935F47"><w:rPr><w:lang w:val="en-US"/></w:rPr></w:r><w:r><w:rPr><w:lang w:val="en-US"/></w:rPr></w:r><w:r w:rsidR="00695252"><w:rPr><w:lang w:val="en-US"/></w:rPr></w:r><w:r><w:rPr><w:lang w:val="en-US"/></w:rPr></w:r></w:p><w:p w14:paraId="4C17C315" w14:textId="51A581D6" w:rsidR="00695252" w:rsidRDefault="00695252" w:rsidP="00695252"><w:pPr><w:rPr><w:lang w:val="en-US"/></w:rPr></w:pPr><w:r w:rsidRPr="00695252"><w:rPr><w:lang w:val="en-US"/></w:rPr><w:t xml:space="preserve">some</w:t></w:r><w:r><w:rPr><w:lang w:val="en-US"/></w:rPr></w:r><w:r w:rsidRPr="00695252"><w:rPr><w:lang w:val="en-US"/></w:rPr></w:r></w:p><w:p w14:paraId="306F6BD7" w14:textId="77777777" w:rsidR="007824CF" w:rsidRDefault="007824CF" w:rsidP="007824CF"><w:pPr><w:rPr><w:lang w:val="en-US"/></w:rPr></w:pPr></w:p><w:p w14:paraId="193F9BCE" w14:textId="39E7FA17" w:rsidR="006E722B" w:rsidRDefault="006E722B" w:rsidP="00F13143"><w:pPr><w:rPr><w:lang w:val="en-US"/></w:rPr></w:pPr><w:r><w:rPr><w:lang w:val="en-US"/></w:rPr><w:t>----</w:t></w:r></w:p><w:p w14:paraId="79276D44" w14:textId="4F382710" w:rsidR="007824CF" w:rsidRPr="007824CF" w:rsidRDefault="007824CF"><w:pPr><w:rPr><w:lang w:val="en-US"/></w:rPr></w:pPr></w:p><w:sectPr w:rsidR="007824CF" w:rsidRPr="007824CF"><w:pgSz w:w="11906" w:h="16838"/><w:pgMar w:top="1440" w:right="1440" w:bottom="1440" w:left="1440" w:header="708" w:footer="708" w:gutter="0"/><w:cols w:space="708"/><w:docGrid w:linePitch="360"/></w:sectPr></w:body></w:document>`;
